ext/dom/php_dom.c

```diff
@@ -1863,38 +2315,323 @@ static zval *dom_nodemap_read_dimension(zend_object *object, zval *offset, int t
 
 	php_dom_named_node_map_get_item_into_zval(php_dom_obj_from_obj(object)->ptr, lval, rv);
 	return rv;
-} /* }}} end dom_nodemap_read_dimension */
+}
 
 static int dom_nodemap_has_dimension(zend_object *object, zval *member, int check_empty)
 {
 	ZVAL_DEREF(member);
 
+	/* If it exists, it cannot be empty because nodes aren't empty. */
+	ZEND_IGNORE_VALUE(check_empty);
+
 	zend_long offset;
 	if (dom_nodemap_or_nodelist_process_offset_as_named(member, &offset)) {
 		/* exceptional case, switch to named lookup */
-		return php_dom_named_node_map_get_named_item(php_dom_obj_from_obj(object)->ptr, Z_STRVAL_P(member), false) != NULL;
+		return php_dom_named_node_map_get_named_item(php_dom_obj_from_obj(object)->ptr, Z_STR_P(member), false) != NULL;
 	}
 
 	return offset >= 0 && offset < php_dom_get_namednodemap_length(php_dom_obj_from_obj(object));
-} /* }}} end dom_nodemap_has_dimension */
+}
 
-xmlNodePtr dom_clone_node(xmlNodePtr node, xmlDocPtr doc, const dom_object *intern, bool recursive)
+static zval *dom_modern_nodemap_read_dimension(zend_object *object, zval *offset, int type, zval *rv)
 {
-	/* See http://www.xmlsoft.org/html/libxml-tree.html#xmlDocCopyNode for meaning of values */
-	int extended_recursive = recursive;
-	if (!recursive && node->type == XML_ELEMENT_NODE) {
-		extended_recursive = 2;
+	if (UNEXPECTED(!offset)) {
+		zend_throw_error(NULL, "Cannot append to %s", ZSTR_VAL(object->ce->name));
+		return NULL;
 	}
-	xmlNodePtr copy = xmlDocCopyNode(node, doc, extended_recursive);
-	if (UNEXPECTED(!copy)) {
+
+	ZVAL_DEREF(offset);
+	if (Z_TYPE_P(offset) == IS_STRING) {
+		php_dom_named_node_map_get_named_item_into_zval(php_dom_obj_from_obj(object)->ptr, Z_STR_P(offset), rv);
+	} else if (Z_TYPE_P(offset) == IS_LONG) {
```

Comment:
I forgot how this is handled for numeric strings?

Reply:
Old DOM did wacky stuff, but did handle numeric strings.
I'm not handling numeric strings here. I don't know if it should?

Reply:
Since a numeric string has to start with a sign character or [0-9] and a name production in XML/HTML cannot start with such a character, I could support it without creating interpretation conflicts.

Reply:
Ok implemented that now for both named maps and node lists.